internal/client/consensus/grandpa/environment.go

```diff
@@ -625,3 +630,1018 @@ type votePrecommit[H runtime.Hash, N runtime.Number] struct {
 }
 
 func (votePrecommit[H, N]) isVote() {}
+
+// The environment we run GRANDPA in.
+type environment[
+	H runtime.Hash,
+	N runtime.Number,
+	Hasher runtime.Hasher[H],
+	Header runtime.Header[N, H],
+	E runtime.Extrinsic,
+] struct {
+	Client        ClientForGrandpa[H, N, Hasher, Header, E]
+	SelectChain   common.SelectChain[H, N, Header]
+	Voters        grandpa.VoterSet[primitives.AuthorityID]
+	Config        Config
+	AuthoritySet  SharedAuthoritySet[H, N]
+	Network       networkBridge[H, N, Hasher]
+	SetID         SetID
+	VoterSetState SharedVoterSetState[H, N]
+	VotingRule    VotingRule[H, N, Header]
+	// TODO: metrics
+	JustificationSender *GrandpaJustificationSender[H, N, Header]
+	// TODO: telemetry
+}
+
+// Updates the voter set state using the given closure. The write lock is held during evaluation of the closure and
+// the environment's voter set state is set to its result if successful.
+func (e *environment[H, N, Hasher, Header, E]) updateVoterSetState(
+	f func(voterSetState voterSetState[H, N]) (voterSetState[H, N], error),
+) error {
+	e.VoterSetState.innerMtx.Lock()
+	defer e.VoterSetState.innerMtx.Unlock()
+
+	newState, err := f(e.VoterSetState.inner)
+	if err != nil {
+		return err
+	}
+	if newState != nil {
+		e.VoterSetState.inner = newState
+	}
+
+	// TODO: metrics
+	return nil
+}
+
+// Report the given equivocation to the GRANDPA runtime module. This method generates a session membership proof of
+// the offender and then submits an extrinsic to report the equivocation. In particular, the session membership proof
+// must be generated at the block at which the given set was active which isn't necessarily the best block if there
+// are pending authority set changes.
+func (e *environment[H, N, Hasher, Header, E]) reportEquivocation(
+	equivocation primitives.Equivocation,
+) error {
+	localID := e.VoterSetState.votingOn(equivocation.Round())
+	if localID != nil {
+		if equivocation.Offender() == *localID {
+			return fmt.Errorf("refraining from sending equivocation report for our own equivocation")
+		}
+	}
+
+	isDescendentOf := utils.IsDescendantOf[H, N, Header](e.Client, nil)
+
+	// TODO [Substrate #9158]: Use SelectChain::best_chain() to get a potentially more accurate best block
```

Comment:
Is this a todo we want here?

Reply:
Yes, I'd like to keep it in until the underlying issue is resolved or they remove their own TODO.  It's a bit messy now since they've moved things to polkadot-sdk so it may not resolve.  I think it would be good for us to contribute back to `polkadot-sdk` when we find these TODOs.